maven_iuvs/fits_processing.py

```diff
@@ -0,0 +1,273 @@
+
+    Returns
+    -------
+    orbit segment as a string 
+
+    """
+    if 'orbit' not in fname:
+        # raise ValueError('IUVS segments only apply to on-orbit data')
+        return "none"
+    
+    seg_pattern = "(?<=iuv_l1[a-c]_)[a-z]+"
+    return re.search(seg_pattern, fname)[0]
+
+
+def locate_missing_frames(hdul, n_int):
+    """
+    Finds missing frames within the FITS HDUList object
+
+    Parameters
+    ----------
+    hdul : astropy FITS HDUList object
+           HDU list for a given observation
+
+    Returns
+    -------
+    n_int : int
+            number of integrations
+
+    """
+    primary_isnan = np.isnan(hdul['PRIMARY'].data)
+    
+    if n_int == 1:
+        # only one integration in file
+        missing_frame = np.any(primary_isnan)
+        if missing_frame:
+            return np.array([0])
+    else:
+        # locate integrations containing NaN values
+        missing_frame_indices = np.where(np.any(primary_isnan,axis=(1,2)))[0]
+        if len(missing_frame_indices) > 0:
+            return missing_frame_indices
+    
+    # no frames missing, data is clean
+    return None
+   
+
+def pix_to_bin(hdul, pix0, pix1, spa_or_spe, return_npix=True):
+    """
+    ...description...
+
+    Parameters
+    ----------
+    hdul : astropy FITS HDUList object
+           HDU list for a given observation
+    pix0 : 
+    pix1 : 
+    spa_or_spe : string
+                 indicates whether this function will convert spatial or 
+                 spectral pixels to bins
+    return_npix : boolean
+                  whether to return the total number of pixels calculated
+
+    Returns
+    -------
+
+    """
+    binpixlo = hdul['Binning'].data[spa_or_spe+'PIXLO'][0]
+    binpixhi = hdul['Binning'].data[spa_or_spe+'PIXHI'][0]
+    binpixwidth = binpixhi+1 - binpixlo
+    
+    nbins = len(binpixlo)
+    
+    binlo = np.searchsorted(binpixlo, pix0+0.01) - 1
+    binlo = 0 if binlo < 0 else binlo
+    binhi = np.searchsorted(binpixhi, pix1-0.01) + 1
+    binhi = nbins if binhi > nbins else binhi
+    
+    if return_npix:
+        npix = np.sum(binpixwidth[binlo:binhi])
+        return binlo, binhi, npix
+    
+    return binlo, binhi
+
+
+def get_pix_range(myfits, which="spatial"):
+    """
+    Given a fits observation, gets the range of pixels
+    for either the spatial or spectral dimension.
+    
+    Parameters
+    ----------
+    myfits : IUVSFITS or HDUList
+             IUVS FITS file in question
+    which: string
+           "spatial" or "spectral"
+           
+    Returns
+    ----------
+    Array of pixel values for bin edges.
+    """
+    pixlo = myfits['Binning'].data[f'{which[:3]}pixlo'][0]
+    pixhi = myfits['Binning'].data[f'{which[:3]}pixhi'][0]
+
+    if not (set((pixhi[:-1]+1)-pixlo[1:]) == {0}):
+        raise ValueError("Error in bin table")
+
+    pixrange = np.concatenate([[pixlo[0]], pixhi+1])
+
+    return pixrange
+
+    
+def update_index(rootpath, new_files_limit_per_run=1000):
```

Comment:
needs to be moved to echelle unless this is going to be made generic for all FUV/MUV files

Reply:
Fixed by 6467166340944ca6a69510b1a6725e519a1a501e